lib/http_aws_sigv4.c

```diff
@@ -66,6 +70,198 @@ static void sha256_to_hex(char *dst, unsigned char *sha, size_t dst_l)
   }
 }
 
+static inline char *find_date_hdr(struct Curl_easy *data, const char *sig_hdr)
+{
+  char *tmp = Curl_checkheaders(data, sig_hdr);
+
+  if(tmp)
+    return tmp;
+  return Curl_checkheaders(data, "Date");
+}
+
+static CURLcode make_headers(struct Curl_easy *data,
+                             const char *hostname,
+                             char *timestamp,
+                             const char *provider1_low,
+                             const char *provider1_mid,
+                             char **date_header,
+                             char **canonical_headers,
+                             char **signed_headers)
+{
+  char *date_hdr_key =  curl_maprintf("X-%s-Date", provider1_mid);
+  char *date_full_hdr = curl_maprintf("x-%s-date:%s",
+                                      provider1_low, timestamp);
+#ifdef DEBUGBUILD
+  char *force_host = getenv("CURL_FORCEHOST");
+#endif
+  struct curl_slist *head = NULL;
+  struct curl_slist *tmp_head = NULL;
+  int ret = CURLE_OUT_OF_MEMORY;
+  struct curl_slist *l;
+  int again = 1;
+
+  if(!date_hdr_key || !date_full_hdr)
+    goto fail;
+
+  if(Curl_checkheaders(data, "Host")) {
+    head = NULL;
+  }
+#ifdef DEBUGBUILD
+  else if(force_host) {
```

Comment:
Just for information. I am not insisting on anything.

Take a look how I overcame the random port issue using `CONNECT_TO` feature:  https://github.com/curl/curl/pull/8202/commits/fc296959cda6ea28d180256d1e68309da7b62ab1
I think this could have advantage that the real code path is tested in this case.